### PR TITLE
Problem: Inefficient binary encoding of Merkle tree path (CRO-149)

### PR DESCRIPTION
**Pull request by yihuang**
I've refactored the ``Path`` representation with ``Vec``, which makes it both more correct and more compact serialization (Old representation allows some illegal path state).
With the current implementation, the gain of further optimizing the serialization with ``BitVec`` is small, I guess it doesn't worth the trouble anymore.

### CODE REVIEW

**Comment by yihuang** [comment]
Sorry about the compile error and format, just fixed them, and made a force push to keep it clean.

**Comment by yihuang** [comment]
I've compared the serialization length between new and old representation, the numbers are below:

| Path length | Old  | New  | Hashes        |
| ----------- | ---- | ---- | ------------- |
| 3           | 168  | 163  | 5*32 = 160    |
| 9           | 570  | 553  | 17 * 32 = 544 |

File: chain-core/src/common/merkle_tree.rs [hunk 1]
```diff
@@ -81,92 +81,97 @@ impl<T> Tree<T> {
...
 
-/// Merkle path for inclusion proof
 #[derive(Debug, PartialEq, Eq, Clone, Encode, Decode)]
-pub struct Path {
+pub struct PathNode {
     node_hash: H256,
-    sibling: Option<Sibling>,
-    sub_path: Option<Box<Path>>,
+    child_hash: H256, // hash of left child unless reversed
+    hash_side: Side,  // the side of hash child
```
Comment (leejw51crypto) [inline chain-core/src/common/merkle_tree.rs]: could add some explanation , which "side" means?

**Comment by leejw51crypto** [comment]
lgtm,
does thi works with the same result of previous one?

**Comment by tomtau** [comment]
> lgtm,
> does this work with the same result of previous one?

yeah, there are a bunch of tests from the old implementation: https://github.com/crypto-com/chain/blob/master/chain-core/src/common/merkle_tree.rs#L381 -- one different thing is that the previous representation could have represented some invalid states, e.g. no Sibling but with subpath or something like that

File: chain-core/src/common/merkle_tree.rs (hunk 1 above)
Comment (yihuang) [reply chain-core/src/common/merkle_tree.rs]: Just added some ascii art to explain it: https://github.com/crypto-com/chain/pull/558/commits/ab63bbc327de13e451a8e43b300a6cb158e77706

**Comment by yihuang** [comment]
done @tomtau

**Comment by tomtau** [comment]
and test compilation fails: https://travis-ci.org/crypto-com/chain/jobs/608661250#L1426
```
failures:
1425
1426---- src/common/merkle_tree.rs - common::merkle_tree::Side (line 121) stdout ----
1427error: expected one of `!`, `.`, `::`, `;`, `?`, `{`, `}`, or an operator, found `,`
1428 --> src/common/merkle_tree.rs:122:11
1429  |
14303 | node_hash1, Left
1431  |           ^ expected one of 8 possible tokens here
1432
1433error: aborting due to previous error
1434
1435Couldn't compile the test.
1436---- src/common/merkle_tree.rs - common::merkle_tree::Side (line 135) stdout ----
1437error[E0422]: cannot find struct, variant or union type `Path` in this scope
1438 --> src/common/merkle_tree.rs:136:1
1439  |
14403 | Path {
1441  | ^^^^ not found in this scope
1442  |
1443help: possible candidate is found in another module, you can import it into scope
1444  |
14452 | use std::path::Path;
1446  |
1447
1448error[E0425]: cannot find value `leaf_hash` in this scope
1449 --> src/common/merkle_tree.rs:137:14
1450  |
14514 |   leaf_hash: leaf_hash,
1452  |              ^^^^^^^^^ not found in this scope
1453
1454error[E0422]: cannot find struct, variant or union type `PathNode` in this scope
1455 --> src/common/merkle_tree.rs:138:12
1456  |
14575 |   nodes: [ PathNode {
1458  |            ^^^^^^^^ not found in this scope
1459
1460error[E0425]: cannot find value `node_hash0` in this scope
1461 --> src/common/merkle_tree.rs:139:16
1462  |
14636 |     node_hash: node_hash0,
1464  |                ^^^^^^^^^^ not found in this scope
1465
1466error[E0425]: cannot find value `child_hash0` in this scope
1467 --> src/common/merkle_tree.rs:140:17
1468  |
14697 |     child_hash: child_hash0,
1470  |                 ^^^^^^^^^^^ not found in this scope
1471
1472error[E0425]: cannot find value `Right` in this scope
1473 --> src/common/merkle_tree.rs:141:16
1474  |
14758 |     hash_side: Right
1476  |                ^^^^^ not found in this scope
1477  |
1478help: possible candidates are found in other modules, you can import them into scope
1479  |
14802 | use core::fmt::Alignment::Right;
1481  |
14822 | use core::fmt::rt::v1::Alignment::Right;
1483  |
14842 | use serde::export::fmt::Alignment::Right;
1485  |
14862 | use serde::export::fmt::rt::v1::Alignment::Right;
1487  |
1488    and 2 other candidates
1489
1490error[E0422]: cannot find struct, variant or union type `PathNode` in this scope
1491 --> src/common/merkle_tree.rs:142:6
1492  |
14939 |   }, PathNode {
1494  |      ^^^^^^^^ not found in this scope
1495
1496error[E0425]: cannot find value `node_hash1` in this scope
1497  --> src/common/merkle_tree.rs:143:16
1498   |
149910 |     node_hash: node_hash1,
1500   |                ^^^^^^^^^^ not found in this scope
1501
1502error[E0425]: cannot find value `child_hash1` in this scope
1503  --> src/common/merkle_tree.rs:144:17
1504   |
150511 |     child_hash: child_hash1,
1506   |                 ^^^^^^^^^^^ not found in this scope
1507
1508error[E0425]: cannot find value `Left` in this scope
1509  --> src/common/merkle_tree.rs:145:16
1510   |
151112 |     hash_side: Left
1512   |                ^^^^ not found in this scope
1513   |
1514help: possible candidates are found in other modules, you can import them into scope
1515   |
15162  | use core::fmt::Alignment::Left;
1517   |
15182  | use core::fmt::rt::v1::Alignment::Left;
1519   |
15202  | use serde::export::fmt::Alignment::Left;
1521   |
15222  | use serde::export::fmt::rt::v1::Alignment::Left;
1523   |
1524     and 2 other candidates
1525
1526error: aborting due to 10 previous errors
1527
1528Some errors have detailed explanations: E0422, E0425.
1529For more information about an error, try `rustc --explain E0422`.
1530Couldn't compile the test.
1531
1532failures:
1533    src/common/merkle_tree.rs - common::merkle_tree::Side (line 121)
1534    src/common/merkle_tree.rs - common::merkle_tree::Side (line 135)
```

**Comment by yihuang** [comment]
Clippy and doctest problem fixed.

**Comment by tomtau** [comment]
bors retry

**Pull request merged**